### PR TITLE
federated learning 2.0

### PR DESCRIPTION
**Pull request by justin1121**

### CODE REVIEW

File: examples/federated-learning/run.py [hunk 1]
```diff
@@ -131,74 +80,124 @@ class DataOwner:
...
+  validation_step(model_owner, player_name=model_owner.player_name)
 
 if __name__ == "__main__":
 
   logging.basicConfig(level=logging.DEBUG)
 
-  model_owner = ModelOwner("model-owner")
+  model = tf.keras.Sequential((
+      tf.keras.layers.Dense(512, input_shape=[None, 28 * 28],
+                            activation='sigmoid'),
```
Comment (yanndupis) [inline examples/federated-learning/run.py]: the activation function should be a `relu` instead of `sigmoid`. thanks

File: examples/federated-learning/convert.py [hunk 2]
```diff
@@ -57,9 +61,10 @@ def normalize(image, label):
 
 def get_data_from_tfrecord(filename, batch_size: int):
```
Comment (mortendahl) [inline examples/federated-learning/convert.py]: is thing function used at all? if not it might be better to simply remove it

File: tf_encrypted/protocol/pond/pond.py [hunk 3]
```diff
@@ -693,6 +693,9 @@ def reconstruct_input(x):
         if isinstance(outputs, tf.Operation):
           return outputs
 
+        if isinstance(outputs, tf.data.Dataset):
```
[the remark this answers is not in the document]
Comment (justin1121) [reply tf_encrypted/protocol/pond/pond.py]: Yeah I don't think this is going to work. Was looking at tensorboard after I submitted the PR.

File: examples/federated-learning/run.py [hunk 4]
```diff
@@ -131,74 +80,124 @@ class DataOwner:
...
-    dataset = tf.data.TFRecordDataset([self.local_data_file])
-    dataset = dataset.map(decode)
-    dataset = dataset.map(normalize)
-    dataset = dataset.repeat()
-    dataset = dataset.batch(self.BATCH_SIZE)
+  This will be performed on the ModelOwner device
+  after securely aggregating gradients.
+
+  Args:
+    *grads: `tf.Variables` representing the federally computed gradients.
```
[the remark this answers is not in the document]
Comment (justin1121) [reply examples/federated-learning/run.py]: I didn't write this comment. Could be stale.

File: examples/federated-learning/run.py [hunk 5]
```diff
@@ -12,11 +14,44 @@
...
+tfe.set_protocol(tfe.protocol.Pond())
+
+EPOCHS = 1
+BATCH_SIZE = 256
+BATCHES = 60000 // BATCH_SIZE
+
+TRACING = True
+
+@tfe.local_computation
+def build_data_pipeline(validation=False, batch_size=BATCH_SIZE):
```
[the remark this answers is not in the document]
Comment (justin1121) [reply examples/federated-learning/run.py]: Nope it doesn't build the graph that way. If you put `build_data_pipeline` in a loop the dataset actually gets built each iteration and returns the same data each iteration.

File: examples/federated-learning/run.py [hunk 6]
```diff
@@ -131,74 +80,124 @@ class DataOwner:
...
+
+    self.dataset = iter(build_data_pipeline(player_name=self.player_name))
 
-  def _build_data_pipeline(self):
-    """Build local data pipeline for federated DataOwners."""
-    def normalize(image, label):
-      image = tf.cast(image, tf.float32) / 255.0
-      return image, label
+@tfe.local_computation
+def update_model(model_owner, *grads):
```
[the remark this answers is not in the document]
Comment (justin1121) [reply examples/federated-learning/run.py]: @jvmancuso @yanndupis suggested that these methods should not live on the classes because we were trying to make it more natural how a data scientist would actually program a training workflow. Personally, I'm not sure it makes a difference and putting the calls on the classes makes the calls a little less verbose by removing `player_name=` part of it.

File: examples/federated-learning/run.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mortendahl) [reply examples/federated-learning/run.py]: that happens even if it's wrapped as a `tf.function`? alternatively you might consider using `tf.device` directly instead of `local_computation`

File: examples/federated-learning/run.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mortendahl) [reply examples/federated-learning/run.py]: maybe it's worth a shot. there's less point in having the `DataOwner` and `ModelOwner` classes in this case though. for the model owner the `player_name=` could be removed by passing it directly to `local_computation`.

i'm wondering if `local_computation` is the right construct to use actually. one issue is that it's protocol dependent, which is why you had to modify Pond, although we should only care about Pond for the secure aggregation part. maybe there's a way to unity these uses of `local_computation`, but also wondering if it wouldn't be just as good simply to use eg `with tf.device(model_owner.device_name):` here?

File: examples/federated-learning/run.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (justin1121) [reply examples/federated-learning/run.py]: I can double check but I'm pretty sure this makes sense because `tf.function` builds the graph in the same order the code is written and called. It's not doing anything unexpected like how the old graph building code worked its doing exactly what you ask it to.

[the remark this answers is not in the document]
Comment (justin1121) [reply examples/federated-learning/run.py]: Also good I idea about using tf.device directly. I'll try it out.

[the remark this answers is not in the document]
Comment (mortendahl) [reply examples/federated-learning/run.py]: in the old way we could get a tensor that would contain a new value for each session.run, while only constructing the data graph once. wondered if tf.function could be used for the same here since it will only build the graph once.

**Comment by justin1121** [comment]
Thanks for the feedback. My intention with this is to only have it work on 2.0.

File: examples/federated-learning/run.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (justin1121) [reply examples/federated-learning/run.py]: Yeah maybe you're right. I'll give it a try too.

File: examples/federated-learning/run.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (justin1121) [reply examples/federated-learning/run.py]: You need pond for securing the grads when they come from the data owner too don't you? And the updating the grads on the model owner's model?

[the remark this answers is not in the document]
Comment (mortendahl) [reply examples/federated-learning/run.py]: > You need pond for securing the grads when they come from the data owner too don't you?

Absolutely! Sharing should happen with Pond.

>  And updating the grads on the model owner's model needs to be secured too?

The aggregated gradients should be revealed to the model owner, but the actual updating could/should happen independently of Pond.

File: examples/federated-learning/convert.py (hunk 2 above)
Comment (justin1121) [reply examples/federated-learning/convert.py]: Deleted!

File: examples/federated-learning/run.py [hunk 7]
```diff
@@ -131,74 +80,124 @@ class DataOwner:
...
   data_owners = [
-      DataOwner("data-owner-0", "./data/train.tfrecord",
-                model_owner.build_update_step),
-      DataOwner("data-owner-1", "./data/train.tfrecord",
-                model_owner.build_update_step),
-      DataOwner("data-owner-2", "./data/train.tfrecord",
-                model_owner.build_update_step),
+      DataOwner("data-owner-0", "./data/train.tfrecord", loss),
+      DataOwner("data-owner-1", "./data/train.tfrecord", loss),
+      DataOwner("data-owner-2", "./data/train.tfrecord", loss),
```
Comment (yanndupis) [inline examples/federated-learning/run.py]: To validate that the model is learning currently, I would suggest that the `data-owner` 0, 1 and 2 are the 'model-owner' are using different dataset (currently always the same `train.tfrecord`). These files could be generated by `download.py`. When generating this dataset, we just need to make sure the MNIST data has been shuffled, in order to avoid that the data-owner-0 got all the 0 and 1 digits, data-owner-1 all the the 1 and 2. thanks

File: examples/federated-learning/run.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (yanndupis) [reply examples/federated-learning/run.py]: If we think we will be always leaking data because of eager execution, maybe we could say that eager execution is just for simulation/experimentation. For a tutorial, it's going to be a lot easier to experiment with eager. 
cc @jvmancuso

[the remark this answers is not in the document]
Comment (justin1121) [reply examples/federated-learning/run.py]: You can experiment using `tf.function` the same as you can with eager execution.

[the remark this answers is not in the document]
Comment (yanndupis) [reply examples/federated-learning/run.py]: good point!

**Comment by jvmancuso** [comment]
Summary of the last commit:
- Tried to enable both higher-level training via `fit` and custom training loops
- the Owner metaclass & the base class for BaseModelOwner handle orchestration
- the user is expected to provide the following:
    - a `model_fn` -- determines how each party comes up with it's model update.
    - an `aggregator_fn` -- determines how the ModelOwner aggregates the results of all DataOwners' model updates.
    - an `evaluator_fn` -- determines how the ModelOwner will evaluate the final model (right now this has little to know specification other than that it exists)
- I'd also like to standardize the build_data_pipeline for DataOwners in a similar way but haven't gotten there yet
- I used the functions from the previous commit as off-the-shelf choices for these three. that should be generic enough to support a decent variety of models & use cases, and we can continue to build these out into a library of commonly-used federated algorithms
    - as an example, I implemented a model_fn for federated [Reptile](https://arxiv.org/abs/1803.02999), a first order meta-learning algorithm that should perform better than federated averaging when parties have relatively non-I.I.D datasets

Currently the model_fn  can be implemented either on the ModelOwner or DataOwner class -- the former makes sense if the model owner is just driving everything, the latter potentially allows different parties to provide different models (e.g. ensembles that need to be trained/aggregated jointly). If we want to force this to be on the DataOwner that works too, but I thought it'd be nice to leave it up to the user.

There's still a lot of work to be done to get this functional, but hopefully this works to unblock @justin1121 and give some further direction. Happy to help get it up and running tomorrow.

File: examples/federated-learning/func_lib.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
+import tensorflow as tf
+import tf_encrypted as tfe
+
+
+### Example model_fns ###
+
+def default_model_fn(data_owner):
```
Comment (mortendahl) [inline examples/federated-learning/func_lib.py]: `model_fn` to me is what constructs the model, as [here in TFF](https://www.tensorflow.org/federated/federated_learning#converters_for_keras)

File: examples/federated-learning/func_lib.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+  # When paired with the secure_aggregation aggregator_fn, this model_fn
+  # corresponds to using g_k as the outer gradient update. See the Reptile
+  # paper for more: https://arxiv.org/abs/1803.02999
+  for k in range(iterations):
+    grads_k = grad_fn(data_owner, **kwargs)
+    data_owner.optimizer.apply_gradients(
+        zip(grads_k, data_owner.model.trainable_variables),
+    )
+
+  return model.trainable_variables
```
Comment (mortendahl) [inline examples/federated-learning/func_lib.py]: just making sure that it's okay to use weights as gradients here..?

File: examples/federated-learning/func_lib.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+
+### Example aggregator_fns ###
+
+def secure_aggregation(collected_inputs):
```
Comment (mortendahl) [inline examples/federated-learning/func_lib.py]: `secure_mean`?

File: examples/federated-learning/run.py [hunk 11]
```diff
@@ -73,116 +75,82 @@ def normalize(image, label):
...
-                       a local federated learning update.
-  """
-  def __init__(self, player_name, local_data_file, model, loss):
-    self.player_name = player_name
-    self.local_data_file = local_data_file
-    self.loss = loss
+### Owner Metaclass ###
 
-    device_name = tfe.get_config().get_player(player_name).device_name
+class Owner(type):
```
Comment (mortendahl) [inline examples/federated-learning/run.py]: duplicate in `players.py`; leftover?

File: examples/federated-learning/players.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+    # device and compiled
+    if hasattr(owner_obj, "aggregator_fn"):
+      self.handle_tf_local_fn(owner_obj, "aggregator_fn")
+
+    if hasattr(owner_obj, "evaluator_fn"):
+      self.handle_tf_local_fn(owner_obj, "evaluator_fn")
+
+    # Decorate user-defined TFE local_computations
+    if has_attr(obj, "model_fn"):
+      self.handle_tfe_local_fn(owner_obj, "model_fn")
```
Comment (mortendahl) [inline examples/federated-learning/players.py]: this smells bad:
- behavior being redefined in base class, away from the user
- base class referring to methods only defined in subclasses

i suggest getting rid of `Owner` as follows:
- turn `pin_to_owner` into a global decorator
- introduce decorated `_call_agggregator_fn`, `_call_evaluator_fn`, and `_call_model_fn`
- use the above in `_runner` and `_update_one_round`

File: examples/federated-learning/func_lib.py (hunk 8 above)
Comment (jvmancuso) [reply examples/federated-learning/func_lib.py]: I was referring to how tf.Estimator refers to it: 
>The Estimator object wraps a model which is specified by a model_fn, which, given inputs and a number of other parameters, returns the ops necessary to perform training, evaluation, or predictions.

Happy to change the name though!

File: examples/federated-learning/func_lib.py (hunk 9 above)
Comment (jvmancuso) [reply examples/federated-learning/func_lib.py]: See the last paragraph in Section 3 of the [Reptile paper](https://arxiv.org/abs/1803.02999), which points to Algorithm 3 in [this NeurIPS paper](https://papers.nips.cc/paper/4006-parallelized-stochastic-gradient-descent) as the preferred way to parallelize Reptile.

Now that you raise this point though, there should be a step where we subtract the averaged weights from the current master model's weights, so this is definitely a bug in the current implementation.

File: examples/federated-learning/players.py (hunk 12 above)
Comment (jvmancuso) [reply examples/federated-learning/players.py]: Yeah, that seems right! The benefit of using a metaclass is that it would enable this behavior in any class created from itself, but it's not much of a benefit here since we know the only two player types we want.

Comment (jvmancuso) [reply examples/federated-learning/players.py]: But I'd say that in general the two points you mentioned are not necessarily bad when emphasized to the user -- tf.Variable does the first, and torch.nn.Module does both. Both of these expect to be subclassed in a variety of different ways, and so behavior is modified so as to make those subclasses comply with extra behavior inside of contexts and the like.

**Comment by justin1121** [comment]
This is good to go on my end again! More feedback welcome.
